docs/decisions/0023-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,218 @@
+  - Activity tags & Metrics
+    - Use activities to record individual operations.
+    - Use tags in activities to record custom information.
+    - Use baggage in activities to more easily correlate dependencies.
+    - Use metrics to continuously monitor performance.
+    - Use tags in metrics to categorize metrics.
+  - Logging & Metrics
+    - Use logging to record custom information for individual operations.
+    - Use metrics to continuously monitor performance.
+    - Use tags in metrics to categorize metrics.
+
+## Decision Outcome
+
+TBD
+
+## Validation
+
+Tests can be added to make sure that all the expected telemetry items are in place and of the correct format.
+
+## Description the Options
+
+### Function hooks
+
+Function hooks allow developers to inject logic to kernel that will be executed before or after a function is invoked. Example use cases include logging the function input before a function is invoked, and logging results after the function returns.
+For more information, please refer to the following ADRs:
+
+1. [Kernel Hooks Phase 1](./0005-kernel-hooks-phase1.md)
+2. [Kernel Hooks Phase 2](./0018-kernel-hooks-phase2.md)
+
+This approach will automatically inject default callbacks that sends telemetry data to a telemetry service for all functions when the feature is turned on.
+
+Pros:
+
+1. Maximum exposure and flexibility to the developers. i.e. App developers can very easily what additional information is needed for individual functions by adding more callbacks.
+
+Cons:
+
+1. Does not use the existing System.Diagnostic framework that has been set up in the kernel. i.e. InstrumentedSKFunction, InstrumentedPlan, and InstrumentedPlanner.
+2. Does not provide the full trace detail.
+3. Difficult to create a telemetry service agnostic solution.
+
+> Note: With distributed tracing and function hook already set up in SK, it's up to the developers to log additional information as telemetry, provided that the information is available in the hooks.
+
+### Distributed tracing
+
+Distributed tracing is a diagnostic technique that can localize failures and performance bottlenecks within distributed applications. .Net has native support to add distributed tracing in your libraries and .Net libraries are also instrumented to produce distributed tracing information automatically.
+
+For more information, please refer to this document: [.Net distributed tracing](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/)
+
+Overall pros:
+
+1. Native .Net support and the kernel has set this up.
+2. Telemetry service agnostic.
+
+Overall cons:
+
+1. Less flexibility for app developers to add custom traces and metrics.
+
+#### Metrics
+
+Pros:
+
+1. Easy to use for continuous monitoring.
+2. Has been set up in SK.
+
+Cons: No additional cons identified given the scope of this document.
+
+#### Activity tags & baggage
+
+```csharp
+// In the instrumented planner
+activity?.AddBaggage("Operation", $"{PlannerType}.CreatePlan");
```

Comment:
Are there unique IDs associated with each operation?

Reply:
We are no longer considering baggage as it will raise security concerns. Updated in the latest version.